### PR TITLE
Workaround a bug that prevents deallocating GPU memory from CPU

### PR DESCRIPTION
**Pull request by e-kayrakli**
This works around a bug that @ShreyasKhandekar discovered while working on Arkouda.

When a GPU-allocated piece of memory is stored in a CPU-based data structure like a `map`, that GPU-allocated memory needs to be freed by the CPU (outside of the GPU sublocale that it was initially allocated on). Our runtime implementation was not ready for that. This PR puts a bandaid with hopes to keep that work going in the short term.

Test:
- [x] ~the arkouda case works fine~ we currently need this fix for something else and it helps
- [x] simpler reproducer on nvidia
- [x] simpler reproducer on amd 
- [x] full nvidia
- [x] full amd

### CODE REVIEW

File: runtime/src/gpu/nvidia/gpu-nvidia.c [hunk 1]
```diff
@@ -352,6 +352,14 @@ void* chpl_gpu_impl_mem_alloc(size_t size) {
 void chpl_gpu_impl_mem_free(void* memAlloc) {
   if (memAlloc != NULL) {
     assert(chpl_gpu_is_device_ptr(memAlloc));
+
+    // see note in chpl_gpu_mem_free
+    int64_t dev_id = c_sublocid_none;
+    CUDA_CALL(cuPointerGetAttribute((void*)&dev_id,
+                                    CU_POINTER_ATTRIBUTE_DEVICE_ORDINAL,
+                                    (CUdeviceptr)memAlloc));
+    switch_context(dev_id);
```
Comment (ShreyasKhandekar) [inline runtime/src/gpu/nvidia/gpu-nvidia.c:361]: I'm assuming that when you switch context to a context that is already active this is a no-op right?

**Review by ShreyasKhandekar — APPROVED** [review]
Looks great! Thanks for figuring this out! And I'll take the award for fastest review ever :P

File: runtime/src/gpu/nvidia/gpu-nvidia.c (hunk 1 above)
Comment (e-kayrakli) [reply runtime/src/gpu/nvidia/gpu-nvidia.c:361]: I don't have a general answer. For NVIDIA, we make sure to not do anything ourselves. So, there's a check involved which could count as an operation: https://github.com/chapel-lang/chapel/blob/main/runtime/src/gpu/nvidia/gpu-nvidia.c#L85-L89

For AMD: we simply set the device as we don't do much with contexts on the AMD side (for the better, I think): https://github.com/chapel-lang/chapel/blob/main/runtime/src/gpu/amd/gpu-amd.c#L69 I would certainly hope that `hipSetDevice` is relatively no-op-y for the scenario you ask about.

Comment (ShreyasKhandekar) [reply runtime/src/gpu/nvidia/gpu-nvidia.c:361]: Okay from some [references](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html), it looks like it should have minimal overhead for duplicate calls. To my brain it initially seemed that trying to switch contexts a lot might be super costly. Thanks for checking!

**Pull request merged**